### PR TITLE
Feature/user authentication

### PR DESCRIPTION
**Pull request by qqmbr4k**

### CODE REVIEW

File: rest-server/server/src/main/resources/reference.conf [hunk 1]
```diff
@@ -10,3 +10,8 @@ rest-server {
   port = 9090
   ssl = false
 }
+
+auth {
+  enabled = false
```
Comment (pferrel) [inline rest-server/server/src/main/resources/reference.conf:15]: Cool.

When we create these compiled in setting we also have an environment variable to override them. So far all env is set or checked here:  https://github.com/actionml/harness/blob/master/rest-server/bin/harness-env

File: rest-server/server/src/main/scala/com/actionml/security/directives/AuthDirectives.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    } else provide(None)
+  }
+
+  def authorizeUser(role: Role, resourceId: ResourceId)
+                   (implicit as: ActorSystem,
+                    mat: ActorMaterializer,
+                    ec: ExecutionContext,
+                    secretOpt: Option[Secret],
+                    log: LoggingAdapter): Directive0 = {
+    if (config.auth.enabled) {
```
Comment (pferrel) [inline rest-server/server/src/main/scala/com/actionml/security/directives/AuthDirectives.scala]: We aren't that picky about style except to follow the scala style docs here: http://docs.scala-lang.org/, which I believe would have these params indented one indent. This is Java style.

File: rest-server/server/src/main/scala/com/actionml/security/model/User.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.actionml.security.model
+
+
+case class User(credentials: Secret,
+                role: Role,
+                resourceId: ResourceId)
```
Comment (pferrel) [inline rest-server/server/src/main/scala/com/actionml/security/model/User.scala]: Scala style is 2 indents for class params.

**Comment by pferrel** [comment]
Looks good so far.

**Comment by pferrel** [comment]
@andreyshikov I will update the drawing here: https://docs.google.com/drawings/d/1_uPiP5UGkphF62PsdIM0Zu8Ugbp4XAKnLYP5CyKrfLo/edit

To include the offline authorization grant where we give toikens to users. Just to cover all the OAuth2 control flow. Correct me if I have it wrong.

File: rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala [hunk 4]
```diff
@@ -3,7 +3,7 @@ package com.actionml.authserver.routes
 import akka.http.scaladsl.model.headers.{BasicHttpCredentials, HttpChallenges}
 import akka.http.scaladsl.server.AuthenticationFailedRejection.{CredentialsMissing, CredentialsRejected}
 import akka.http.scaladsl.server._
-import com.actionml.authserver.model.Client
+import com.actionml.authserver.model.{Client, ClientId}
```
Comment (pferrel) [inline rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala]: I've taken to calling this "Users" in the docs.

Comment (andreyshikov) [reply rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala]: It's not user, it's client - the user's app server

Comment (pferrel) [reply rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala]: The word "client" means too many things. You are talking about an ID for something like a user account, right? Not a client to some server.

Comment (andreyshikov) [reply rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala]: No, it's about app server - same as at oauth2 spec

Comment (pferrel) [reply rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala]: Why did you want to call these "Users" I put that into the REST API URIs POST /users/...

Comment (pferrel) [reply rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala]: Can you check the new full REST API spec to make sure it is correct? Here: https://github.com/actionml/harness/blob/master/rest_spec.md

Note the use of `/users/<user-id>` in the Harness Rest-server API. Is the user-id what you are calling the `ClientId`?

Comment (pferrel) [reply rest-server/auth-server/src/main/scala/com/actionml/authserver/routes/AuthorizationController.scala]: I'd rather leave the REST URIs referencing User collections, it doesn't have to match what you call it in the code neccessarily

**Pull request merged**